components/libwebsockets/examples/client/main/lws-client.c

```diff
@@ -0,0 +1,310 @@
+        lws_close_reason(wsi, LWS_CLOSE_STATUS_NORMAL, (unsigned char *)"bye", 3);
+        /* Return non-null to close the connection. */
+        return -1;
+        break;
+    default:
+        break;
+    }
+
+    return lws_callback_http_dummy(wsi, reason, user, in, len);
+}
+
+static const struct lws_protocols protocols[] = {
+    {
+        "lws-echo",
+        callback_minimal_echo,
+        1024, 1024, 0, NULL, 0
+    },
+    LWS_PROTOCOL_LIST_TERM
+};
+
+int app_main(void)
+{
+    ESP_LOGI(TAG, "[APP] Startup..");
+    ESP_LOGI(TAG, "[APP] Free memory: %" PRIu32 " bytes", esp_get_free_heap_size());
+    ESP_LOGI(TAG, "[APP] IDF version: %s", esp_get_idf_version());
+    esp_log_level_set("*", ESP_LOG_INFO);
+
+    ESP_ERROR_CHECK(nvs_flash_init());
+    ESP_ERROR_CHECK(esp_netif_init());
+    ESP_ERROR_CHECK(esp_event_loop_create_default());
+
+    /* This helper function configures Wi-Fi or Ethernet, as selected in menuconfig.
+     * Read "Establishing Wi-Fi or Ethernet Connection" section in
+     * examples/protocols/README.md for more information about this function.
+     */
+    ESP_ERROR_CHECK(example_connect());
+
+    /* Configure WDT. */
+    TaskHandle_t handle = xTaskGetCurrentTaskHandle();
+    esp_task_wdt_add(handle);
+
+    /* Create LWS Context - Client. */
+    struct lws_context_creation_info info;
+    int n = 0, logs = LLL_USER | LLL_ERR | LLL_WARN | LLL_NOTICE;
+
+    memset(msg, 'x', sizeof(msg));
+
+    lws_set_log_level(logs, NULL);
+    ESP_LOGI(TAG, "LWS minimal ws client echo\n");
+
+    memset(&info, 0, sizeof info); /* otherwise uninitialized garbage */
+    info.port = CONTEXT_PORT_NO_LISTEN; /* we do not run any server */
+    info.protocols = protocols;
+    info.fd_limit_per_thread = 1 + 1 + 1;
+
+#if CONFIG_WS_OVER_TLS_MUTUAL_AUTH
+    info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT;
+
+    /* Configuring client certificates for mutual authentification */
+    extern const char cert_start[] asm("_binary_client_cert_pem_start"); // Client certificate
+    extern const char cert_end[]   asm("_binary_client_cert_pem_end");
+    extern const char key_start[] asm("_binary_client_key_pem_start"); // Client private key
+    extern const char key_end[]   asm("_binary_client_key_pem_end");
+
+    info.client_ssl_cert_mem        = cert_start;
+    info.client_ssl_cert_mem_len    = cert_end - cert_start;
+    info.client_ssl_key_mem         = key_start;
+    info.client_ssl_key_mem_len     = key_end - key_start;
+#elif CONFIG_WS_OVER_TLS_SERVER_AUTH
+    info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT;
+
+    extern const char cacert_start[] asm("_binary_ca_cert_pem_start"); // CA certificate
+    extern const char cacert_end[] asm("_binary_ca_cert_pem_end");
+
+    info.client_ssl_ca_mem          = cacert_start;
+    info.client_ssl_ca_mem_len      = cacert_end - cacert_start;
+#endif
+
+    context = lws_create_context(&info);
+    if (!context) {
+        ESP_LOGE(TAG, "lws init failed\n");
```

Comment:
nit: you don't have to append newline (`\n') to ESP_LOGs
```suggestion
        ESP_LOGE(TAG, "lws init failed");
```